File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/util/Debouncer.java
package io.casey.musikcube.remote.util;

import android.os.Handler;
import android.os.Looper;

public abstract class Debouncer<T> {
    private Handler handler = new Handler(Looper.getMainLooper());
    private T t = null;
    private long delay;

    public Debouncer(long delay) {
        this.delay = delay;
    }

    public void call() {
        handler.removeCallbacks(deferred);
        handler.postDelayed(deferred, delay);
    }

    public void call(T t) {
        this.t = t;
        handler.removeCallbacks(deferred);
        handler.postDelayed(deferred, delay);
    }

    public void cancel() {
        handler.removeCallbacks(deferred);
    }

    protected abstract void onDebounced(T caller);

    private Runnable deferred = () -> {
        onDebounced(t);
    };
}

File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/util/Debouncer.kt
package io.casey.musikcube.remote.util

import android.os.Handler
import android.os.Looper

abstract class Debouncer<in T>(private val delay: Long) {
    private val handler = Handler(Looper.getMainLooper())
    private var last: T? = null

    fun call() {
        handler.removeCallbacks(deferred)
        handler.postDelayed(deferred, delay)
    }

    fun call(context: T) {
        last = context
        handler.removeCallbacks(deferred)
        handler.postDelayed(deferred, delay)
    }

    fun cancel() {
        handler.removeCallbacks(deferred)
    }

    protected abstract fun onDebounced(last: T?)

    private val deferred:Runnable = object : Runnable {
        override fun run() {
            onDebounced(last)
        }
    }
}
